src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -240,9 +247,81 @@ public void Dispose()
             {
                 if (this.toRelease != null)
                 {
-                    this.toRelease.semaphore.Release();
+                    this.toRelease.Release();
                 }
             }
         }
+
+        private static void CancellationHandler(object? state)
+        {
+            var waiterInfo = (WaiterInfo)state!;
+
+            // The party that manages to complete or cancel the task is responsible to remove it from the queue.
+            if (waiterInfo.Trigger.TrySetCanceled(waiterInfo.CancellationToken))
+            {
+                // If the node is in the queue, remove it.
+                // It might not have been added yet if cancellation was already requested by the time we called Register.
+                lock (waiterInfo.Owner.syncObject)
+                {
+                    if (waiterInfo.Owner.waiters.Find(waiterInfo) is { } node)
+                    {
+                        waiterInfo.Owner.waiters.Remove(node);
+                        waiterInfo.Owner.RecycleNode(node);
+                    }
+                }
+            }
+
+            // Clear registration and references.
+            waiterInfo.Cleanup();
+        }
+
+        private void RecycleNode(LinkedListNode<WaiterInfo> node)
+        {
+            Assumes.True(Monitor.IsEntered(this.syncObject));
+            if (this.nodePool.Count < 10)
```

Comment:
Would it make sense to base this off the processor count or whatever the thread pool or Task infrastructure uses to balance itself?

Reply:
I doubt the TPL or ThreadPool classes expose any API to help me reuse their choices for such things. 
This isn't the concurrency count but rather the expected *wait* count, so I'm not sure basing it on processor count makes sense. I imagine it will vary by application more than processor count. For example someone who doesn't expect to ever have more than 1 or 2 requests for a semaphore based on how they use it won't need 32 recyclable nodes on a 32 core processor and that would just waste memory. But someone even on a single core machine might run an app that uses this semaphore like a queue and throw hundreds of requests at it in bursts over time, making the case for a very large recyclable collection here.
So 10 here is just a guess that is small enough to not be a memory hog, but large enough to reduce GC pressure in most common cases.
If an application that uses this class finds this is an intolerable number, we can always expose it as a property so they can fiddle with it.